/*
 * Copyright (c) 2022-2023 NVIDIA CORPORATION & AFFILIATES, ALL RIGHTS RESERVED.
 *
 * This software product is a proprietary product of NVIDIA CORPORATION &
 * AFFILIATES (the "Company") and all right, title, and interest in and to the
 * software product, including all associated intellectual property rights, are
 * and shall remain exclusively with the Company.
 *
 * This software product is governed by the End User License Agreement
 * provided with the software product.
 *
 */

syntax = "proto3";

/* Network port for Firefly */
enum eNetworkPort
{
	k_Dummy = 0;

	/* Default network port */
	k_DocaFirefly = 25600;
};

/* Host (x86) -> DPU (Arm):
 * ========================
 * gRPC server that allows users to subscribe to monitor events.
 */
service FireflyMonitor {
	/* Subscribe for monitoring events */
	rpc Subscribe (SubscribeReq) returns (stream MonitorRecord);
}

/* Here for future compatibility */
message SubscribeReq {
}

/* PTP Stability state */
enum ePTPState
{
	k_STATE_STABLE = 0;				/* PTP is in a stable state */
	k_STATE_FAULTY = 1;				/* PTP is currently out of sync */
	k_STATE_RECOVERED = 2;				/* PTP managed to recover from a sync error */
};

message SamplingValue {
	int64 max = 1;					/* Maximal (abs) sampled value  */
	int64 average = 2;				/* Average value across samples */
}

message MonitorRecord {
	bool gm_present = 1;				/* Is the Grandmaster present? */
	ePTPState ptp_stability = 2;			/* Stability state of the PTP */
	string ptp_time = 3;				/* The accurate timestamp */
	string sys_time = 4;				/* The system's time in a string format */
	uint32 error_count = 5;				/* Number of errors we encountered thus far */
	string last_error_time = 6;			/* Timestamp for the last encountered error */
	string gm_identity = 7;				/* Grandmaster identity */
	string port_identity = 17;			/* Port identity */
	SamplingValue master_offset = 8;		/* Offset from the master clock (in nano seconds) */
	int64 current_utc_offset = 9;			/* Current offset from UTC */
	bool time_traceable = 10;			/* PTP timeTraceable property */
	bool frequency_traceable = 11;			/* PTP frequencyTraceable property */
	uint32 grandmaster_priority1 = 12;		/* Priority1 field of the Grandmaster clock */
	uint32 gm_clock_class = 13;			/* Clock class of the Grandmaster clock */
	uint32 gm_clock_accuracy = 14;			/* Clock accuracy of the Grandmaster clock */
	uint32 grandmaster_priority2 = 15;		/* Priority2 field of the Grandmaster clock */
	uint32 gm_offset_scaled_log_variance = 16;	/* Offset scaled log variance of the Grandmaster clock */
}
